cmd/traffic/cmd/manager/internal/ambassadoragent/cloudtoken/service.go

```diff
@@ -0,0 +1,121 @@
+	cloudConnectTokenKey = "CLOUD_CONNECT_TOKEN"
+)
+
+// NewCreateConfigmapIfNotPresent will start up watchers to watch for a configmap or secret
+// with the apikey. If one is found, the watchers shut down. If MaybeAddToken is called
+// and successfully creates a configmap, the watchers are shut down.
+func NewCreateConfigmapIfNotPresent(ctx context.Context) *createConfigmapIfNotPresent {
+	managerns := managerutil.GetEnv(ctx).ManagerNamespace
+	clientset := k8sapi.GetK8sInterface(ctx)
+	watchers := newTokenWatchers(clientset, managerns)
+
+	// ctx for watchers, cancelled if token is found
+	cancelCtx, cancel := context.WithCancel(ctx)
+
+	go func() {
+		// search checks if apikey token exists, if it does, shut down the watchers
+		search := func() {
+			configmaps, err := watchers.mapsWatcher.List(cancelCtx)
+			if err != nil {
+				dlog.Errorf(ctx, "error watching cloud token configmaps: %s", err.Error())
+				cancel()
+			}
+			for _, cm := range configmaps {
+				_, ok := cm.Data[cloudConnectTokenKey]
+				if strings.HasSuffix(cm.Name, suffix) && ok {
+					dlog.Infof(ctx, "configmap %s found, stopping cloud token watchers", cm.Name)
+					cancel()
+				}
+			}
+			secrets, err := watchers.secretsWatcher.List(cancelCtx)
+			if err != nil {
+				dlog.Errorf(ctx, "error watching cloud token secrets: %s", err.Error())
+				cancel()
+			}
+			for _, s := range secrets {
+				_, ok := s.Data[cloudConnectTokenKey]
+				if strings.HasSuffix(s.Name, suffix) && ok {
+					dlog.Infof(ctx, "secret %s found, stopping cloud token watchers", s.Name)
+					cancel()
+				}
+			}
+		}
+
+		// startup and search
+		dlog.Info(ctx, "starting cloud token watchers")
+		search()
+
+		// context for subscribe. This has to be different than the watcher context
+		// because cancelling the subscribe ctx will close the channel, and a case
+		// will activate on a closed channel. we dont want search triggered after
+		// watchers are shut down
+		subCtx, subCancel := context.WithCancel(ctx)
+
+		// search on broadcast
+		for {
+			select {
+			case <-watchers.subscribe(subCtx):
+				search()
+			case <-cancelCtx.Done():
+				subCancel()
+				return
+			}
+		}
+	}()
+
+	return &createConfigmapIfNotPresent{
+		createToken: func(ctx context.Context, apikey string) error {
+			dlog.Info(ctx, "patching cloud token configmap with apikey")
+			_, err := clientset.CoreV1().ConfigMaps(managerns).Patch(
+				ctx,
+				"traffic-manager-"+suffix,
+				types.StrategicMergePatchType,
+				[]byte(fmt.Sprintf(`{"data":{"%s":"%s"}}`, cloudConnectTokenKey, apikey)),
+				apiv1.PatchOptions{},
+			)
+			if err == nil {
+				dlog.Info(ctx, "cloud token configmap successfully patched")
+			}
+			return err
+		},
+		done: cancelCtx.Done(),
+	}
+}
+
+// MaybeAddToken will add a token if one does not already exist
+func (c *createConfigmapIfNotPresent) MaybeAddToken(ctx context.Context, apikey string) error {
+	select {
+	case <-c.done: // apikey is already present or watcher err, do nothing
```

Comment:
What happens if the apiKey is removed from the config map after an API key has been found (or if the secret is removed )? Seems to me like the `createToken` will never be called unless the traffic-manager is restarted.

Reply:
Meh. If we need to manage the token, we have to keep both watchers alive indefinitely, which is ok but i dont like it. Imo, if someone deletes the apikey, they can add it again too. If they really want the traffic manager to do it, they can roll the traffic manager

Reply:
I don't understand why the watcher is a one-shot thing. If we can assume that the traffic manager can be rolled out when the key is re-added, why can't we assume that the key has been added prior to it being installed and skip the watcher altogether?

In my view, we either assume that it's there, or we assume that it it's controlled by someone else (it can be created, updated, or deleted). Just catering for the fact that it may exist already or arrive after the traffic-manager boots up is, IMO, a somewhat strange middleground. What happens if the token expires and is updated?

Reply:
I dont think the a-agent cares about expired tokens, it opens the connection and never closes it. I can ask the run team if expired tokens have been a problem.

I dont want to step on emissary apikeys because emissary requires a user to manually set the apikey, meaning it is likely an org key, and the incoming login might be a personal key, which would reduce functionality.